src/algorithms/correlators.jl

```diff
@@ -0,0 +1,140 @@
+    @assert all(c₁ .< cs) "The first column index must be less than the second."
+
+    (Nr, Nc) = size(ket)
+    corr = T[]
+
+    left_start = start_left(
+        env.edges[4, mod1(r, Nr), _prev(c₁, Nc)],
+        env.corners[1, _prev(r, Nr), _prev(c₁, Nc)],
+        env.corners[4, _next(r, Nr), _prev(c₁, Nc)],
+    )
+    left_side =
+        left_start * MPSKit.TransferMatrix(
+            env.edges[1, _prev(r, Nr), mod1(c₁, Nc)],
+            (ket[mod1(r, Nr), mod1(c₁, Nc)], O[1], bra[mod1(r, Nr), mod1(c₁, Nc)]),
+            _dag(env.edges[3, _next(r, Nr), mod1(c₁, Nc)]),
+        )
+    left_side_norm =
+        left_start * MPSKit.TransferMatrix(
+            env.edges[1, _prev(r, Nr), mod1(c₁, Nc)],
+            (ket[mod1(r, Nr), mod1(c₁, Nc)], bra[mod1(r, Nr), mod1(c₁, Nc)]),
+            _dag(env.edges[3, _next(r, Nr), mod1(c₁, Nc)]),
+        )
+    for c in (c₁ + 1):cs[end]
+        if c ∈ cs
+            left_side_final =
+                left_side * MPSKit.TransferMatrix(
+                    env.edges[1, _prev(r, Nr), mod1(c, Nc)],
+                    (ket[mod1(r, Nr), mod1(c, Nc)], O[2], bra[mod1(r, Nr), mod1(c, Nc)]),
+                    _dag(env.edges[3, _next(r, Nr), mod1(c, Nc)]),
+                )
+            final = end_right(
+                left_side_final,
+                env.edges[2, mod1(r, Nr), _next(c, Nc)],
+                env.corners[2, _prev(r, Nr), _next(c, Nc)],
+                env.corners[3, _next(r, Nr), _next(c, Nc)],
+            )
+        end
+        (left_side, left_side_norm) = [
+            l * MPSKit.TransferMatrix(
+                env.edges[1, _prev(r, Nr), mod1(c, Nc)],
+                (ket[mod1(r, Nr), mod1(c, Nc)], bra[mod1(r, Nr), mod1(c, Nc)]),
+                _dag(env.edges[3, _next(r, Nr), mod1(c, Nc)]),
+            ) for l in (left_side, left_side_norm)
+        ]
+        if c ∈ cs
+            final_norm = end_right(
+                left_side_norm,
+                env.edges[2, mod1(r, Nr), _next(c, Nc)],
+                env.corners[2, _prev(r, Nr), _next(c, Nc)],
+                env.corners[3, _next(r, Nr), _next(c, Nc)],
+            )
+            push!(corr, final / final_norm)
+        end
+    end
+    return corr
+end
+
+function correlator_horizontal(
+    bra::InfinitePEPS,
+    O::AbstractTensorMap{T,S,2,2},
+    i::CartesianIndex{2},
+    j::CartesianIndices{2},
+    ket::InfinitePEPS,
+    env::CTMRGEnv,
+) where {T,S}
+    U, Σ, V = tsvd(O, ((1, 3), (2, 4)))
+    O₁ = permute(U * sqrt(Σ), ((1,), (2, 3)))
+    O₂ = permute(sqrt(Σ) * V, ((1, 2), (3,)))
+    return correlator_horizontal(bra, (O₁, O₂), i, j, ket, env)
+end
+
+function correlator_vertical(
+    bra::InfinitePEPS,
+    O,
+    i::CartesianIndex{2},
+    j::CartesianIndices{2},
+    ket::InfinitePEPS,
+    env::CTMRGEnv,
+)
+    i_rot = CartesianIndex(i[2], i[1])
+    j_rot = CartesianIndex(j[1][2], j[1][1]):CartesianIndex(j[end][2], j[end][1])
```

Comment:
This looks like it will not produce correct results if the initial range did not have step size one I fear